### PR TITLE
FloorspaceJS to OpenStudio reverse translator. FIrst implementation p…

### PR DESCRIPTION
**Pull request by ggartside**
…assing unit tests.

Pull request overview
---------------------

 - Fixes #ISSUENUMBERHERE (IF THIS IS A DEFECT)

### Pull Request Author

 - [ ] Model API Changes / Additions
 - [ ] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [ ] Model API methods are tested (in `src/model/test`)
 - [ ] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources: Add Link
 - [ ] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [ ] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [ ] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

**Comment by ggartside** [comment]
This is a floorspace to osm reverse translator. It consists of two distinct modules, a module to to deserialize the floorspace json file into memory and a module to convert the floorspace model into an osm model.

The floorspace module comprises a set of classes that each represent concepts in the floorspace model and the structure of the floorspace model. These classes also implement the visitor design pattern via `Accept `- `Dispatch `methods as defined in the visitor interface, `FSDispatchable`.

The reverse translator is implemented in two parts, a public facing class implements the main API method, `modelFromFloorspace`, and a private visitor class that is implemented solely in the cpp file. Each Dispatch method in the visitor is used where appropriate to create the OSM objects form the floorspace objects.

For example the `Dispatch `method for `FSStory` calls the `Accept` method on all the spaces, the `FSSpace` `Dispatch` method Creates the OpenStudio Space and the surfaces and sub-surfaces and similarly for all the building shading elements too.

**Review by jmarrec — CHANGES_REQUESTED** [review]
I haven't done an in-depth review of the actual functionality yet. But there are issues in the design IMHO. I have made some of changes as I went through the review and pushed to this branch but there are more to address.

The core question I have if whether the visitor pattern is really needed in that case. I may be missing the larger/long term picture here, but currently it seems to bring complexity for no benefit since there's no actual double dispatching happening. Interface classes may be enough and simpler.

----

@ggartside I've started to resolve some of the comments myself because I pushed some fixes in order to remove some clobber, but you should expand and read them anyways (some you may disagree with). Thanks!

File: src/utilities/floorspace/FSModel.hpp [hunk 1]
```diff
@@ -0,0 +1,568 @@
...
+
+namespace openstudio {
+
+// forward declaration
+class Point3d;
+class Vector3d;
+class FSModel;
+
+// Base class for all floorspace model items
+class UTILITIES_API FSBase : FSDispatchable
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/utilities/floorspace/FSModel.hpp:55]: I have done some pragma ignores in 35bf7ba1475724a276f21319a77de3a88b2be558 for the time being, but this needs to be addressed / justified.

File: src/utilities/floorspace/FSModel.cpp [hunk 2]
```diff
@@ -0,0 +1,1302 @@
...
+***********************************************************************************************************************/
+
+#include "FSModel.hpp"
+
+#include "../core/Assert.hpp"
+#include "../core/Json.hpp"
+#include "../geometry/point3d.hpp"
+#include "../geometry/vector3d.hpp"
+#include "../geometry/Geometry.hpp"
+#include "../geometry/intersection.hpp"
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/utilities/floorspace/FSModel.cpp]: Done in 924972a8a80857e96ce24ca4f2e440637c6bb4b7

File: src/model/FloorspaceReverseTranslator.cpp [hunk 3]
```diff
@@ -0,0 +1,998 @@
...
+    ABOVECEILING
+  };
+
+  // TODO: Creating surfaces for spaces and creatig shading surfaces for building shading have
+  // a lot of code in common but there are significant differences also. At some point we
+  // shoud look into creating some common functions that calculate the vertices
+
+  // A Visitor design pattern class. Each Dispatch method handles creation of OSM
+  // concepts from the FloorspaceJS concepts
+  class FloorspaceReverseTranslator_Impl : public FSVisitor
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/model/FloorspaceReverseTranslator.cpp]: done in f570bf50e66cc18c72e2959ad65b1871263e14a3

File: src/utilities/floorspace/FSModel.cpp [hunk 4]
```diff
@@ -0,0 +1,1302 @@
...
+
+void FSFace::clearEdgeRefs() {
+  m_edgeRefs.clear();
+}
+
+void FSFace::setEdgeRefs(std::vector<FSEdgeReference> edgeRefs) {
+  for (auto& edgeRef : edgeRefs) {
+    m_edgeRefs.push_back(edgeRef);
+  }
+}
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/utilities/floorspace/FSModel.cpp]: Done in aa721bf60938f95ed76bf97b1c5b10ce747929d1

File: src/utilities/floorspace/FSModel.cpp [hunk 5]
```diff
@@ -0,0 +1,1302 @@
...
+    throw openstudio::Exception(std::string("Cannot find key '" + key + "'"));
+  }
+}
+
+// assert type is correct if key is present
+void FSModel::assertType(const Json::Value& value, const std::string& key, const Json::ValueType& valueType) {
+  if (!checkType(value, key, valueType)) {
+    throw openstudio::Exception(std::string("Key '" + key + "' is of wrong type"));
+  }
+}
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/utilities/floorspace/FSModel.cpp]: Done in bfa1cc5297cd0349810030d38b2265d1a10e7157

File: src/utilities/floorspace/FSModel.cpp [hunk 6]
```diff
@@ -0,0 +1,1302 @@
...
+
+void FSModel::Accept(FSVisitor& visitor) const {
+  visitor.Dispatch(*this);
+}
+
+//std::ostream& operator<<(std::ostream& os, const Plane& plane) {
+//  os << "[" << plane.a() << ", " << plane.b() << ", " << plane.c() << ", " << plane.d() << "]";
+//  return os;
+//}
+//
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/utilities/floorspace/FSModel.cpp]: Done in bfa1cc5297cd0349810030d38b2265d1a10e7157

File: src/utilities/floorspace/FSModel.cpp [hunk 7]
```diff
@@ -0,0 +1,1302 @@
...
+#include "FSModel.hpp"
+
+#include "../core/Assert.hpp"
+#include "../core/Json.hpp"
+#include "../geometry/point3d.hpp"
+#include "../geometry/vector3d.hpp"
+#include "../geometry/Geometry.hpp"
+#include "../geometry/intersection.hpp"
+
+#include <iomanip>
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/utilities/floorspace/FSModel.cpp]: Done in e7a48b2c8ab5deb6b459bd4e62c51af11901125b

File: src/utilities/floorspace/FSModel.hpp [hunk 8]
```diff
@@ -0,0 +1,568 @@
...
+
+  std::vector<FSThermalZone> thermalZones() const;
+  std::vector<FSBuildingUnit> buildingUnits() const;
+  std::vector<FSBuildingType> buildingTypes() const;
+  std::vector<FSSpaceType> spaceTypes() const;
+  std::vector<FSConstructionSet> constructionSets() const;
+  std::vector<FSStory> stories() const;
+  const std::vector<FSDaylightingControlDefinition> daylightingControlDefinitions() const;
+  const std::vector<FSWindowDefinition> windowDefinitions() const;
+  const std::vector<FSDoorDefinition> doorDefinitions() const;
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/utilities/floorspace/FSModel.hpp]: Done in bfa1cc5297cd0349810030d38b2265d1a10e7157

File: src/utilities/floorspace/FSModel.hpp [hunk 9]
```diff
@@ -0,0 +1,568 @@
...
+class UTILITIES_API FSConfig
+{
+ public:
+  FSConfig();
+  FSConfig(const Json::Value& root);
+
+  std::string units();
+  std::string language();
+  bool unitsEditable();
+  double northAxis() const;
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/utilities/floorspace/FSModel.hpp]: done in bfa1cc5297cd0349810030d38b2265d1a10e7157

File: src/utilities/floorspace/FSModel.hpp [hunk 10]
```diff
@@ -0,0 +1,568 @@
...
+};
+
+class UTILITIES_API FSStory : public FSBase
+{
+ public:
+  FSStory();
+  FSStory(const Json::Value& root, const FSModel& model);
+  void Accept(FSVisitor& visitor) const;
+
+  double gelowFloorPlenumHeight() const;
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/utilities/floorspace/FSModel.hpp]: Done in 0dc2f6b6810f2abb5f260b686055a000fb5ba36c

File: src/utilities/floorspace/FSModel.cpp [hunk 11]
```diff
@@ -0,0 +1,1302 @@
...
+    return m_edge.secondVertex();
+}
+
+///////////////////////////////////////////////////////////////////////////////////
+
+FSFiller::FSFiller() {
+  m_alpha = 0.5;
+}
+
+FSFiller::FSFiller(const Json::Value& root, const FSModel& model, FSStory& story) : FSBase(root) {
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/utilities/floorspace/FSModel.cpp]: Done in bfa1cc5297cd0349810030d38b2265d1a10e7157

File: src/utilities/floorspace/FSModel.cpp [hunk 12]
```diff
@@ -0,0 +1,1302 @@
...
+
+boost::optional<FSDoorDefinition> FSDoor::doorDefinition() const {
+  return m_doorDefinition;
+}
+
+///////////////////////////////////////////////////////////////////////////////////
+
+FSShading::FSShading() {}
+
+FSShading::FSShading(const Json::Value& root, const FSModel& model, FSStory& story) : FSBase(root) {
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/utilities/floorspace/FSModel.cpp]: Done in bfa1cc5297cd0349810030d38b2265d1a10e7157

File: src/utilities/floorspace/FSModel.cpp [hunk 13]
```diff
@@ -0,0 +1,1302 @@
...
+
+      // Establishes the unit conversion from the project's config property.
+      setUnits();
+
+      // Load all the space assignment objects so that when the space is loaded
+      // we can find the objects via their ids
+      const Json::Value& thermalZones = m_value.get("thermal_zones", Json::arrayValue);
+      for (Json::ArrayIndex t = 0; t < thermalZones.size(); t++) {
+        m_thermalZones.emplace_back(FSThermalZone(thermalZones[t]));
+      }
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/utilities/floorspace/FSModel.cpp]: Done in bfa1cc5297cd0349810030d38b2265d1a10e7157

File: src/utilities/floorspace/FSModel.cpp [hunk 14]
```diff
@@ -0,0 +1,1302 @@
...
+      m_project.load(project, *this);
+
+      // Establishes the unit conversion from the project's config property.
+      setUnits();
+
+      // Load all the space assignment objects so that when the space is loaded
+      // we can find the objects via their ids
+      const Json::Value& thermalZones = m_value.get("thermal_zones", Json::arrayValue);
+      for (Json::ArrayIndex t = 0; t < thermalZones.size(); t++) {
+        m_thermalZones.emplace_back(FSThermalZone(thermalZones[t]));
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/utilities/floorspace/FSModel.cpp]: done in bfa1cc5297cd0349810030d38b2265d1a10e7157

File: src/utilities/floorspace/FSModel.cpp [hunk 15]
```diff
@@ -59,53 +60,53 @@ void FSModel::load(const std::string& json) {
       formattedErrors.clear();
       parsingSuccessful = Json::parseFromStream(rbuilder, ifs, &m_value, &formattedErrors);
     }
+  }
 
-    if (parsingSuccessful) {
+  if (parsingSuccessful) {
```
Comment (jmarrec) [inline src/utilities/floorspace/FSModel.cpp:65]: `} else {`

File: src/model/test/FloorspaceReverseTranslator_GTest.cpp [hunk 16]
```diff
@@ -88,70 +88,70 @@ void CompareTwoModels(Model& model, Model& baseline) {
   // Compare stories
   //   Match by name,compare properties
   //   Compare shading count
-
+  double tol = 0.001;  // Same tolerance as defaulyts for circularEqual
```
Comment (jmarrec) [inline src/model/test/FloorspaceReverseTranslator_GTest.cpp:91]:
```suggestion
  double tol = 0.001;  // Same tolerance as defaults for circularEqual
```

**Comment by DavidGoldwasser** [comment]
With latest installer, fenestration by WWR seems to work when running the measure unit test from model articulation gem
https://github.com/NREL/openstudio-model-articulation-gem/compare/develop...floorspace_js_translation

Screen Shot 2022-08-08 at 11 26 37 AM

**Review by DavidGoldwasser — APPROVED** [review]

**Comment by tijcolem** [comment]
@DavidGoldwasser Thanks for confirming.  This looks good to me. Ignore the CI on the last commit as it was failing due to BCL issues unrelated to the this PR.

**Pull request merged**